### PR TITLE
support cardinality iseq

### PR DESCRIPTION
**Pull request by bbrehm**
This adds support for properties that are indexedSeq (in practice: ArraySeq).

### CODE REVIEW

**Comment by fabsx00** [comment]
Should we call it `array` rather than `iseq`? Sounds like a concept that is more widely known.

File: src/main/scala/overflowdb/codegen/CodeGen.scala
Comment (ml86) [inline src/main/scala/overflowdb/codegen/CodeGen.scala:740]: Please stop using fully qualified names. It is just ugly and makes reading harder. If you are just writing `immutable.IndexedSeq` and put and import up top everybody knows what you are talking about.

**Review by ml86 — APPROVED** [review]

**Review by mpollmeier — APPROVED** [review]

**Comment by bbrehm** [comment]
>Should we call it array rather than iseq? Sounds like a concept that is more widely known.

Well, three issues:

1. Array is super nice for performance, but really kinda sucks in scala (very easy to get the beloved "missing classtag" error, bad variance behavior, counter-intuitive iseqal). When looping over the array, then some loop invariant code motion pass in the JIT should hoist the extra indirection, so that's not too bad -- so the only real performance price is that we potentially have one more L1 miss when we first access the array. This is avoided by recreating the wrapper on-demand, which has a good chance of being elided (scalar replacement of aggregates JIT optimization). 

There is an argument for storing the wrapped array instead of storing the raw array and recreating the wrapper on demand. In general, an extra allocation is cheaper than an L1 miss -- but if the JIT is too stupid to hoist / scalar replace, and we sometimes fail to manually hoist the accessor out of loops, then these allocations add up and storing the wrapper is preferable. If anyone here knows more about JVM performace than me, please tell me which way is better. @hubertp @tuxology ?

2. We really don't want people to think they can modify the result!

3. Array is a concrete final type, not an interface. In other words, we cannot change the representation without breaking compatibility. By using IndexedSeq we can decide in the future to change the type. For example, we may consider implementing our own array-wrapper that supports fast tail operations (the wrapper stores both an offset and a pointer to the underlying array) or extends `Anyval` or caches the hash of its entries (useful if we want to use `flow.points` as key inside a set / hashmap).

**Comment by fabsx00** [comment]
Ah, I did not mean using the scala type for an array, I just meant using the name `array` in the specification as opposed to `iseq` - and still translating to indexed seq.

File: src/main/scala/overflowdb/codegen/CodeGen.scala
Comment (hubertp) [inline src/main/scala/overflowdb/codegen/CodeGen.scala:459]:
```suggestion
          case (Cardinality.List | Cardinality.ISeq, _) => ""
```

File: src/main/scala/overflowdb/codegen/CodeGen.scala
Comment (hubertp) [inline src/main/scala/overflowdb/codegen/CodeGen.scala:851]: please don't write empty match errors

File: src/main/scala/overflowdb/codegen/CodeGen.scala
Comment (hubertp) [inline src/main/scala/overflowdb/codegen/CodeGen.scala:907]: same as above

File: src/main/scala/overflowdb/codegen/Helpers.scala
Comment (hubertp) [inline src/main/scala/overflowdb/codegen/Helpers.scala:27]: shouldn't this be `ISeq` as well?

**Review by hubertp — COMMENTED** [review]
I would agree with @fabsx00 that `ISeq` is a bit of a new term. better stick to known short names or use full name instead

File: src/main/scala/overflowdb/codegen/Helpers.scala
Comment (bbrehm) [reply src/main/scala/overflowdb/codegen/Helpers.scala:27]: The HigherValueType representation is only used for edge properties -- and we don't need to support array-valued edge properties.

This is not just lazyness: Either we use and test `array`/`iseq` cardinality for at least one edge property, or the code will rot away.

**Comment by bbrehm** [comment]
So, to update, apart from cosmetics (use `"array"` in the json):
1. I talked myself into storing the wrapped object, instead of the raw Array
2. Upon further consideration, we currently only support the new `array` cardinality for containedNode-style pseudo-properties. This is good because of code-rot: We only use and test it for such pseudo-properties, so it's better to throw in codegen than to emit possibly faulty/untested code

I'll merge once markus approves the newest revision.

**Comment by ml86** [comment]
I liked the raw object better, but can live with both.

**Review by ml86 — APPROVED** [review]

**Pull request merged**
